src/EditorFeatures/Core/BackgroundWorkIndicator/BackgroundWorkIndicatorContext.cs

```diff
@@ -89,212 +51,116 @@ public BackgroundWorkIndicatorContext(
+            lock (this.ContextAndScopeDataMutationGate)
             {
-                Contract.ThrowIfFalse(_scopes.Contains(scope));
-                _scopes = _scopes.Remove(scope);
+                Contract.ThrowIfFalse(_scopes_onlyAccessUnderLock.Contains(scope));
+                _scopes_onlyAccessUnderLock = _scopes_onlyAccessUnderLock.Remove(scope);
             }
 
-            // We changed.  Enqueue work to make sure the UI reflects this.
-            this.EnqueueUIUpdate();
+            ReportTotalProgress();
         }
 
-        private (string description, ProgressInfo progressInfo) BuildData()
+        private void ReportTotalProgress()
         {
-            lock (Gate)
-            {
-                var description = _firstDescription;
-                var progressInfo = new ProgressInfo();
+            // Lightup the UI if it supports IProgress
+            if (_backgroundWorkIndicator is IProgress<ProgressInfo> underlyingProgress)
+                underlyingProgress.Report(GetTotalProgress());
 
-                foreach (var scope in _scopes)
+            ProgressInfo GetTotalProgress()
+            {
+                lock (ContextAndScopeDataMutationGate)
                 {
-                    var scopeData = scope.ReadData_MustBeCalledUnderLock();
+                    var progressInfo = new ProgressInfo();
 
-                    // use the description of the last scope if we have one.  We don't have enough room to show all
-                    // the descriptions at once.
-                    description = scopeData.description;
+                    foreach (var scope in _scopes_onlyAccessUnderLock)
+                    {
+                        var scopeProgressInfo = scope.ProgressInfo_OnlyAccessUnderLock;
+                        progressInfo = new ProgressInfo(
+                            progressInfo.CompletedItems + scopeProgressInfo.CompletedItems,
+                            progressInfo.TotalItems + scopeProgressInfo.TotalItems);
+                    }
 
-                    var scopeProgressInfo = scopeData.progressInfo;
-                    progressInfo = new ProgressInfo(
-                        progressInfo.CompletedItems + scopeProgressInfo.CompletedItems,
-                        progressInfo.TotalItems + scopeProgressInfo.TotalItems);
+                    return progressInfo;
                 }
-
-                return (description, progressInfo);
             }
         }
 
-        string IUIThreadOperationContext.Description => BuildData().description;
+        string IUIThreadOperationContext.Description
+        {
+            get
+            {
+                // use the description of the last scope if we have one.  We don't have enough room to show all
+                // the descriptions at once.
+                lock (this.ContextAndScopeDataMutationGate)
+                    return _scopes_onlyAccessUnderLock.LastOrDefault()?.Description ?? _firstDescription;
```

Comment:
>_firstDescription [](http://example.com/codeflow?start=87&length=17)

If the initial scope that was added during construction is removed, then it seems like we shouldn't be returning _firstDescription (which would allow removal of that member)

Reply:
added comment.  basically, there is always an implicit scope the platform holds onto which holds onto that first description.  they just don't expose it.  so this is how we simulate deferring to them.